js/pathfinding.js

```diff
@@ -5,56 +5,250 @@ import {settingsKey} from "./settings.js";
 import {getSnapPointForTokenObj, getTokenSize, iterPairs} from "./util.js";
 
 import * as GridlessPathfinding from "../wasm/gridless_pathfinding.js";
-import {PriorityQueueSet} from "./data_structures.js";
+import {PriorityQueueSet, ProcessOnceQueue} from "./data_structures.js";
 
+/**
+ * Class to hold all the cached node data, and functions to deal with caching.
+ * 
+ * Since pathfinding can depend on several factors, e.g. the token's size, we keep
+ * several caches, keyed by all the data relevant to pathfinding. If we already have
+ * the maximum number of caches and we need to create another one, we discard the
+ * one not used for the longest.
+ * 
+ * When we select a token, or a token we have selected updates, we start caching
+ * in the background so, when we do start pathfinding, it's very performant.
+ * 
+ * Background caching starts by trying to run an idle process (when the browser is
+ * otherwise not busy), but if it can't do that after an amount of time (e.g. the
+ * CPU is very slow and is busy) then we instead start caching a few nodes each
+ * frame. 
+ */
 class Cache {
 	static maxCacheIds = 5;
+	static maxBackgroundCachingMillis = 10;
+	static maxAnimationCachingMillis = 5;
+	static backgroundCachingTimeoutMillis = 200;
 
 	constructor() {
 		this.nodes = new Map();
 		this.lastUsed = new Map();
+		this.background = {
+			queues: new Map(),
```

Comment:
These three maps are all mappings from the cache id to some property of the Cache entry. I think we should consolidate them into a single map that maps to an object containing all three properties.